subsystems/visionSensor.py

```diff
@@ -0,0 +1,78 @@
+import wpilib
+import json
+from lib1076.udp_channel import UDPChannel
+
+class VisionSensor:
+    """
+    Vision sensor has two main roles.   It listens on the specified
+    port for udp angle packets and parses them to obtain the most
+    recent value for the angle to target.  It also acts as a data
+    source for a PIDController.
+
+    If the target is to the left of the robot, the angle is reported
+    as negative.
+
+    Example:
+        visionSensor = VisionSensor('10.10.76.7', 8812)
+
+        This creates channel to listen on port 8812 for angle information.
+    """
+    def __init__(self, sensor_ip, listen_port, simulation=False, logger=None):
+        self.vision_ip = sensor_ip
+        self.vision_port = listen_port
+        self.logger = logger
+        self.simulation = simulation
+        self.bearing = 0
+        self.range_cm = 0
+        self.channel = self.createChannel()
+
+    def pidGetBearing(self):
+        # return cached bearing that was last received from
+        # the vision system
+        return self.bearing
+
+    def pidGetRange(self):
+        # return cached range that was last received from
+        # the vision system
+        return self.range_cm
+
+    def getPIDSourceType(self):
+        return wpilib.interfaces.pidsource.PIDSource.PIDSourceType.kDisplacement
+
+    def createChannel(self):
+        if self.simulation is False:
+            VISION_IP = self.vision_ip
+            VISION_PORT = self.vision_port
+            LOCAL_IP = "10.10.76.2"
+        else:
+            VISION_IP = "127.0.0.1"
+            VISION_PORT = 8812
+            LOCAL_IP = "127.0.0.1"
+        
+        try:
+            channel = UDPChannel(local_ip=LOCAL_IP, local_port=VISION_PORT, 
```

Comment:
This needs to be self.channel

Reply:
Nope.  I got that wrong.   This is local variable to return.  By Bad.  Resolved.